### ISSUE TITLE
Non-public database properties should behave as any database property in SQL 

### ISSUE THREAD

**Issue by k-rus** [Starcounter/Starcounter#2508]
This issue is important for migrating databases from one Starcounter or application version to another. Thus all database properties will be accessible within SQL queries. 
Currently, it is necessary that all properties are public for migration, which, IMHO, is much worse than allowing properties in SQL queries.
This issue is only about `develop`. In `next` a better solution is proposed.

Note that this will provide read-only access to non-public properties, while the reflection already provides read/write access to the properties.

**Comment by per-samuelsson** [comment]
Just a linecheck that I understand.

```C#
[Database] public class Foo {
  public Foo(string bar) { Bar = bar; }
  string Bar { get; set; }
}

class Program {
  static void Main() {
   Db.Transaction(() => {new Foo("Qux");});
   Trace.Assert(Db.SQL<Foo>("SELECT f FROM Foo f WHERE f.Bar == ?", "Qux").First != null);
  }
}
```

This should work and not trigger an assertion failure (in an empty database)?

**Comment by k-rus** [comment]
Yes, this is correct. To my understanding he only thing, which is necessary to do, is to create propertyDef in Binding for non-public properties. This change should solve the issue. Let me know if it doesn't go.

My suggestion is to implement #2509 first to see that it returns correct message. This will allow diagnostic problems when they appear again and help users.

**Comment by k-rus** [comment]
> This should work and not trigger an assertion failure (in an empty database)?

Actually I am not sure if I understood the question correctly. Why do you mean assertion failure?

The following code should work:
```c#
Database] public class Foo {
  private string _Bar;
  public string Bar { get { return _Bar; } set { _Bar = value; } }
}

class Program {
  static void Main() {
   Db.Transaction(() => { 
      var f = new Foo();
      f.Bar = "Qux"; 
    });
   Trace.Assert(Db.SQL<Foo>("SELECT f FROM Foo f WHERE f.Bar == ?", "Qux").First != null);
  }
}
```

**Comment by per-samuelsson** [comment]
> Actually I am not sure if I understood the question correctly. Why do you mean assertion failure?

I meant my assertion should not trigger (i.e. it should be possible to do a select on the private field and get a result back).

Your second example is not at all the same as mine. It does a query on a **public** property, while my query is on the private one (note that I have no public modifier on the Bar, while you do).

**Comment by per-samuelsson** [comment]
> To my understanding he only thing, which is necessary to do, is to create propertyDef in Binding for non-public properties.

I doubt its as easy as that. But I will spend more time on this tomorrow. I'm trying to understand the code right now, and to set up a proper environment (switching from next, etc).

**Comment by per-samuelsson** [comment]
Where is the original issue, the one that describe their problem?

**Comment by per-samuelsson** [comment]
> Where is the original issue, the one that describe their problem?

Found it myself (#2485)

**Comment by eriohl** [comment]
I suspect this will break a lot of code since it seems likely that private and public properties will share the same name is case-insensitive.

For example this:

```c#
Database]
public class Foo {
  private string bar;
  public string Bar { get { return bar; } set { bar = value; } }
}
```

Will no longer work right? (Yes, in this case Bar and bar does the same thing but this doesn't have to be case and so it can't be assumed to be the case).

(Also that code has read/write access to private fields using reflection isn't entirely true I think. At least in .NET 2.0 this was a matter of code access control).

**Comment by per-samuelsson** [comment]
@k-rus 

Am I correct in the assumption that what we are trying to achieve here is to support a one-time migration? Basically, to do this so they can move from one model where they use private fields, into another where they dont?

Otherwise, I can't see how it could be sufficient only with read-only access; the reload will need to write them.

**Comment by per-samuelsson** [comment]
@eriohl also has a valid point. I'm concerned about the same thing. Maybe we can work around it with some "smart" logic, that recognize that pattern in the binding though (or even in the unload, if my above assumption is correct), but it will be a bit ugly. Then again everything with this is a bit ugly anyway :crying_cat_face:

**Comment by per-samuelsson** [comment]
Another question: do we need to support private reference fields as well? Because as has been reported earlier, declaring such fields blow up the code host (since metadata population fail on this). For this, I did the redesign in **next**. Must the same thing work in **develop**?

**Comment by k-rus** [comment]
> I meant my assertion should not trigger (i.e. it should be possible to do a select on the private field and get a result back).

The assertion needs to be switched off. It is never hit in the current implementation, since query parser fails during type checking phase. But unload will fail on it after the change.



> Your second example is not at all the same as mine. It does a query on a public property, while my query is on the private one (note that I have no public modifier on the Bar, while you do).

Sorry, my bad. I meant:
```c#
Database] public class Foo {
  private string _Bar;
  public string Bar { get { return _Bar; } set { _Bar = value; } }
}

class Program {
  static void Main() {
   Db.Transaction(() => { 
      var f = new Foo();
      f.Bar = "Qux"; 
    });
   Trace.Assert(Db.SQL<Foo>("SELECT f FROM Foo f WHERE f._Bar == ?", "Qux").First != null);
  }
}
```


> I suspect this will break a lot of code since it seems likely that private and public properties will share the same name is case-insensitive.
> 
> For example this:
> 
> Database]
> public class Foo {
>   private string bar;
>   public string Bar { get { return bar; } set { bar = value; } }
> }
> Will no longer work right? (Yes, in this case Bar and bar does the same thing but this doesn't have to be case and so it can't be assumed to be the case).

To my understanding it should not brake the execution. Columns and properties are treated separately.



> Am I correct in the assumption that what we are trying to achieve here is to support a one-time migration? Basically, to do this so they can move from one model where they use private fields, into another where they dont?
> 
> Otherwise, I can't see how it could be sufficient only with read-only access; the reload will need to write them.

Reload does not use internal .NET meta-data. It uses meta-tables. Fixing .NET meta-data should be enough to fix the meta-data. No additional work should be needed.



> Another question: do we need to support private reference fields as well? Because as has been reported earlier, declaring such fields blow up the code host (since metadata population fail on this). For this, I did the redesign in next. Must the same thing work in develop?

In my understanding it is enough just to add non-public properties to PropertyDefs. However, I cannot be sure on 100%, since I don't know all the code around it.

**Comment by per-samuelsson** [comment]
> To my understanding it should not brake the execution. Columns and properties are treated separately

But it would mean you would have two properties only differenting in case. Is that allowed? I thought names where case insesitive?

**Comment by k-rus** [comment]
> To my understanding it should not brake the execution. Columns and properties are treated separately
> But it would mean you would have two properties only differenting in case. Is that allowed? I thought names where case insesitive?

Yes, you are right. Both of them will become properties. Thank you pushing this issue and @eriohl for finding it.
Now it is trade of between to evals. I don't know how often the pattern is used in customer code.

**Comment by per-samuelsson** [comment]
> I don't know how often the pattern is used in customer code

More than "never ever" at least, which is what count. Because it mean we must somehow address it. Forcing users to refactor their code prior to **next** to get around it is not an option IMO. Instead, we must handle it.

One idea is that we use a special naming convention for properties we expose that map to private fields, and adapt the unload to detect them whenever it can not find a property mapping for a certain column. This is quite a hack, and it will look ugly in the SQL when doing ```SELECT *```, but then again, it's quite ugly to expose private data in the first place IMO, and it's rather temporary (we are addressing all these issues in **next**).

With the above being suggested, I also still don't know it it would solve the problem, since I don't know how the reload is implemented. I guess the idea is that the dump we are trying to produce (with included private data) must also be possible to reload in an adapted model, i.e. fixing the unload to pass by itself is not enough.

**Comment by per-samuelsson** [comment]
Just for the record, I have reported about exactly this case early december, in #2436 (and related #2061). We should keep that issue updated with what we end up doing here.

**Comment by per-samuelsson** [comment]
> One idea is that we use a special naming convention for properties we expose that map to private fields, and adapt the unload to detect them whenever it can not find a property mapping for a certain column.

> With the above being suggested, I also still don't know if it would solve the problem, since I don't know how the reload is implemented.

Regarding the above two paragraphs, I have now spent a little time to try understand better how the new unload/reload operates. Can I get it confirmed that the below is the basic flow?

**Unload**
1. Get each raw view
2. Fetch every COLUMN
  * 2.1 In the INSERT, use the raw COLUMN NAME
  * 2.2 In the SELECT, use the PROPERTY NAME (property with same name need to be present)

Reload
1. Apply INSERT using column names (any user code level is bypassed)

If I'm right in this, I guess we can (at least in theory) alter **2.2** to try some alternatives in any case it can't find a property with the same name, right? And if so, my initial suggestion would also at least be doable in theory. And furthermore, if we adapt **2.2** not only to look for a property with the **same name**, but also consider **any property mapping to the column**, we could use that to get around this case? (I think I've seen some code in the metadata population that use a similar approach - reaching out for any property mapping to a certain column, to utilize metadata of the property). If we adapt the unload like that, wouldn't it work "just" to require that there was at least one property exposing any column that are to be unloaded? And if so, we could even ask Tunity (or SO, or whoever owns this old model they are trying to get away from), to explicitly add such properties as a one-time thing?

**Comment by per-samuelsson** [comment]
> 2.2 In the SELECT, use the PROPERTY NAME (property with same name need to be present)

Actually, I just realized that I think the unload already try to use properties mapping to a certain column, not the direct property = column name approach. If I'm right in this, can't we just at least get around this temporary by asking customers / SO-team to add read-only properties to any private data to be able to unload?

**Comment by k-rus** [comment]
> More than "never ever" at least, which is what count. Because it mean we must somehow address it. Forcing users to refactor their code prior to next to get around it is not an option IMO. Instead, we must handle it.

We are pushing now to make big refactoring in the users code to replace all non-public properties to be public.


> This is quite a hack, and it will look ugly in the SQL when doing SELECT *, but then again, it's quite ugly to expose private data in the first place IMO

It is much more ugly to require that all properties and fields should be public in the user code. And this is what we require know from our users if they want to have data and keep them while migrating between application or Starcounter version.

> Reload
> 1. Apply INSERT using column names (any user code level is bypassed)

Implementation of INSERT requires information in meta-data, which is not there about columns representing non-public fields/properties. This is a problem, which can be solved by treating all properties in the same way. The only drawback is reported by @eriohl.

**Comment by k-rus** [comment]
I fully agree with @per-samuelsson  suggestion, which we discussed on Hangouts. And this suggestion is confirmed.

**Comment by per-samuelsson** [comment]
> I fully agree with @per-samuelsson suggestion, which we discussed on Hangouts. And this suggestion is confirmed.

After doing a prototype for the solution we agreed upon, I now found it doesn't work. The reason is simple: the generated code bindings require that any PropertyDef is backed by a real .NET PropertyInfo (i.e. the generated binding needs to actually read the, to return a value).

I don't know if its doable to create a binding on the fly, generating code that access the underlying column instead, or if I must rethink the solution. In any case, this will take more time still. :cry:

**Comment by per-samuelsson** [comment]
> I don't know if its doable to create a binding on the fly, generating code that access the underlying column instead.

I considered and studied this a bit, and I don't think it's a way forward. Maybe it's not even doable technically actually, since the weaver emit only private accessors for private data (of course).

With that being said, this unfurtonatelly became a bit larger to solve, if no one can come up with some other solution. As a understand it now, we have to implement something similar to auto-emitted acessor properties in the weaver instead, which is much more time-consuming than the original idea (i.e. adapting the binding). I guess its like 1-2 days of work, once we have first established it can at all be done.

**Comment by per-samuelsson** [comment]
It was decided that we'll just implement a check during binding that refuse to load applications that contain non-public database-bound fields if they are not exposed by at least one public property. The code host will emit an error for each such found entry and contain a reference to a wiki page that describe the action needed by the developer (to expose them at least once).

In **next**, we'll have better support for private data again, by rewriting the unload/reload and adapt metadata.

**Comment by per-samuelsson** [comment]
This issue is now resolved, using the strategy described in the previous comment. It is merged and pushed, along with #2509, in commit https://github.com/Starcounter/Starcounter/commit/00eccd1419c1e5f3d4f8a0ca35923983ce121c66.
